clusterloader2/pkg/measurement/common/pod_startup_latency.go

```diff
@@ -0,0 +1,293 @@
+	runTimes      map[string]metav1.Time
+	watchTimes    map[string]metav1.Time
+}
+
+// Execute supports two actions:
+// - start - Starts to observe pods and pods events.
+// - gather - Gathers and prints current pod latency data.
+func (p *podStartupLatencyMeasurement) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+	var summaries []measurement.Summary
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return summaries, err
+	}
+
+	switch action {
+	case "start":
+		p.namespace, err = util.GetStringOrDefault(config.Params, "namespace", metav1.NamespaceAll)
+		if err != nil {
+			return summaries, err
+		}
+		p.labelSelector, err = util.GetStringOrDefault(config.Params, "labelSelector", "")
+		if err != nil {
+			return summaries, err
+		}
+		p.fieldSelector, err = util.GetStringOrDefault(config.Params, "fieldSelector", "")
+		if err != nil {
+			return summaries, err
+		}
+		return summaries, p.start(config.ClientSet)
+	case "gather":
+		return p.gather(config.ClientSet)
+	default:
+		return summaries, fmt.Errorf("unknown action %v", action)
+	}
+
+}
+
+func (p *podStartupLatencyMeasurement) start(c clientset.Interface) error {
+	glog.Infof("Starting pod startup latency measurement...")
+	p.stopCh = make(chan struct{})
+	optionsModifier := func(options *metav1.ListOptions) {
+		options.FieldSelector = p.fieldSelector
+		options.LabelSelector = p.labelSelector
+	}
+	listerWatcher := cache.NewFilteredListWatchFromClient(c.CoreV1().RESTClient(), "pods", p.namespace, optionsModifier)
+	p.informer = cache.NewSharedInformer(listerWatcher, &corev1.Pod{}, 0)
+	p.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			p.checkPod(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			p.checkPod(newObj)
+		},
+	})
+
+	go p.informer.Run(p.stopCh)
+	timeoutCh := make(chan struct{})
+	timeoutTimer := time.AfterFunc(informerSyncTimeout, func() {
+		close(timeoutCh)
+	})
+	defer timeoutTimer.Stop()
+	if !cache.WaitForCacheSync(timeoutCh, p.informer.HasSynced) {
+		return fmt.Errorf("timed out waiting for caches to sync")
+	}
+	return nil
+}
+
+func (p *podStartupLatencyMeasurement) gather(c clientset.Interface) ([]measurement.Summary, error) {
+	glog.Infof("Gathering pod startup latency measurement...")
+	if p.informer == nil || p.stopCh == nil {
+		return []measurement.Summary{}, fmt.Errorf("metric %s has not been started", podStartupLatencyMetricName)
+	}
+
+	scheduleLag := make([]time.Duration, 0)
+	startupLag := make([]time.Duration, 0)
+	watchLag := make([]time.Duration, 0)
+	schedToWatchLag := make([]time.Duration, 0)
+	e2eLag := make([]time.Duration, 0)
+
+	close(p.stopCh)
+	if err := p.gatherScheduleTimes(c); err != nil {
+		return []measurement.Summary{}, err
+	}
+	for key, create := range p.createTimes {
+		sched, ok := p.scheduleTimes[key]
+		if !ok {
+			glog.Infof("Failed to find schedule time for %v", key)
+			continue
```

Comment:
This is tricky.
If there is pod that stuck in scheduling, starting, being reported etc., this will simply ignore it.
This is wrong - we should ignore them.

Maybe it's fine to not have it as part of histogram (though it's also not obvious to me), but at the very least, we should fail the test (well not immediately, but ensure that the test will fail), if there are more than X% of missed observations.

Reply:
Done.
Measurement will verify if >= 99% pods were successfully scheduled.